web3/_utils/method_formatters.py

```diff
@@ -452,6 +457,16 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
 ABI_REQUEST_FORMATTERS = abi_request_formatters(STANDARD_NORMALIZERS, RPC_ABIS)
 
 
+def raise_invalid_parity_mode(response: RPCResponse) -> NoReturn:
+    error_message = response['error']['message']
```

Comment:
@njgheorghita I'm getting an error from mypy on this line: `error: Invalid index type "str" for "Union[RPCError, str]"; expected type "Union[int, slice]"` and I can't figure out what's wrong with the line. Do you have any ideas?

Reply:
Hmmm, this is a weird one. I'm not sure off the top of my head what's going on.... The types `RPCResponse` and `RPCError` look like they should work just fine with this. Can you try overriding it maybe? with something like `error_message: str = response['error']['message']`

Reply:
hmm, that didn't work either :-/ FWIW, it's green until I add on the `['message']` part

Reply:
And when I do: 
```
def raise_invalid_parity_mode(response: RPCResponse) -> NoReturn:
    error: RPCError = response['error']
    error_message: str = error['message']
    raise InvalidParityMode(error_message)
```
I get: 
`error: Incompatible types in assignment (expression has type "Union[RPCError, str]", variable has type "RPCError")`

Reply:
hmmm, ok i think if I understand [this](https://stackoverflow.com/questions/43982085/mypy-error-incompatible-type-despite-using-union) correctly - with `RPCResponse['error']: Union[RPCError, str]` mypy is throwing a bug b/c the case when `RPCResponse['error']` is a string is not supported (i.e. you can't get the `['message']` key of a string). In this code we're assuming that it will always be an `RPCError` type  -  which is not always the case if I'm remembering correctly (since mypy throws errors in  `web3/providers/eth_tester/main.py` with `error: RPCError`.... but i could be completely wrong on that). 

So, here.. if it's safe to assume that parity will always respond with an `RPCError` type for `response['error']` then i'd just go ahead and 
```python
from typing import cast
def raise_invalid_parity_mode(response: RPCResponse) -> NoReturn:
    error = cast(RPCError, response['error'])
    error_message = error['message']
    raise InvalidParityMode(error_message)
```

Reply:
>mypy is throwing a bug b/c the case when `RPCResponse['error']` is a string is not supported (i.e. you can't get the `['message']` key of a string)

That is correct. `RPCResponse` explicitly defines `error` as being `Union[RPCError, str]` hence it is now rightfully rejects to access an `message` property as that may lead to a runtime error.

My I suggest a different solution than the cast though. Keep in mind a cast always means "step away, `mypy`, I know what I'm doing, please accept `x` to be of type `SomeType`." Mypy will accept anything you feed it even though this may lead to runtime errors.

So in this case, I believe the safer solution would be one of these:

**1. Assuming `string` might be possible**

if isinstance(response.error, RPCError):
   error_message = response.error.message
elif isinstance(response.error, str):
  error_message = response.error
else:
   raise Exception("Invariant") 

**2. Assuming `string` might not be possible** (to be clear the `RPCResponse` type allows for it to be `str`

if isinstance(response.error, RPCError):
   error_message = response.error.message
else:
   raise Exception("Invariant")

The `isinstance` checks allow mypy to ensure type safety which is generally better than giving up type safety by introducing a `cast`.

Reply:
Aha! That makes sense. Thanks @njgheorghita and @cburgdorf!